### PR TITLE
Fix DETA save_pretrained

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

`save_pretrained` is not working for `DetaForObjectDetection`. This PR is supposed to fix this error.

A simple script to reproduce:
```python
from transformers import DetaConfig, DetaForObjectDetection

config = DetaConfig()
model = DetaForObjectDetection(config)
model.save_pretrained("output_dir/")
```

or

```python
from transformers import AutoModelForObjectDetection

model = AutoModelForObjectDetection.from_pretrained("jozhang97/deta-resnet-50")
model.save_pretrained("output_dir/")
```

The following error occurs:

```
RuntimeError: The weights trying to be saved contained shared tensors [{'model.decoder.class_embed.0.weight', 'class_embed.0.weight'}, {'model.decoder.class_embed.0.bias', 'class_embed.0.bias'}, {'model.decoder.class_embed.1.weight', 'class_embed.1.weight'}, {'model.decoder.class_embed.1.bias', 'class_embed.1.bias'}, {'model.decoder.class_embed.2.weight', 'class_embed.2.weight'}, {'model.decoder.class_embed.2.bias', 'class_embed.2.bias'}, {'model.decoder.class_embed.3.weight', 'class_embed.3.weight'}, {'class_embed.3.bias', 'model.decoder.class_embed.3.bias'}, {'model.decoder.class_embed.4.weight', 'class_embed.4.weight'}, {'model.decoder.class_embed.4.bias', 'class_embed.4.bias'}, {'model.decoder.class_embed.5.weight', 'class_embed.5.weight'}, {'class_embed.5.bias', 'model.decoder.class_embed.5.bias'}, {'class_embed.6.weight', 'model.decoder.class_embed.6.weight'}, {'class_embed.6.bias', 'model.decoder.class_embed.6.bias'}] that are mismatching the transformers base configuration. Try saving using `safe_serialization=False` or remove this tensor sharing.
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qubvel** [comment]
cc @NielsRogge

**Review by NielsRogge — APPROVED** [review]
Thanks for fixing!

File: tests/models/deta/test_modeling_deta.py [hunk 1]
```diff
@@ -520,6 +524,43 @@ def test_initialization(self):
...
+                ptrs[id_tensor_storage(tensor)].append(name)
+
+            # These are all the pointers of shared tensors.
+            tied_params = [names for _, names in ptrs.items() if len(names) > 1]
+
+            tied_weight_keys = model_tied._tied_weights_keys if model_tied._tied_weights_keys is not None else []
+            # Detect we get a hit for each key
+            for key in tied_weight_keys:
+                if not any(re.search(key, p) for group in tied_params for p in group):
+                    raise ValueError(f"{key} is not a tied weight key for {model_class}.")
```
Comment (amyeroberts) [inline tests/models/deta/test_modeling_deta.py:550]: We shouldn't really be raising exceptions in a test. Exceptions are for terminating out of code when there's an incorrect input / code behaviour which we can then choose to handle. In tests, we're really performing sanity checks, which should always be True given the test. 

Either we should test this behaviour with an assert or just remove.

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing and adding a test! 

Just a small comment on removing the exception

File: tests/models/deta/test_modeling_deta.py (hunk 1 above)
Comment (qubvel) [reply tests/models/deta/test_modeling_deta.py]: Thanks, Amy, I fixed this test case and the original one too 👍

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating!

**Pull request merged**